service/src/main/java/org/cbioportal/service/impl/UuidDataAccessTokenServiceImpl.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.commons.logging.LogFactory;
+import org.cbioportal.model.DataAccessToken;
+import org.cbioportal.persistence.DataAccessTokenRepository;
+import org.cbioportal.service.DataAccessTokenService;
+import org.cbioportal.service.exception.MaxNumberTokensExceededException;
+import org.cbioportal.service.exception.TokenNotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Service
+@Component("uuid")
+public class UuidDataAccessTokenServiceImpl implements DataAccessTokenService {
+
+    @Autowired
+    private DataAccessTokenRepository dataAccessTokenRepository;
+
+    @Value("${dat.ttl_seconds:-1}")
+    private int datTtlSeconds;
+
+    @Value("${dat.uuid.max_number_per_user:-1}")
+    private int maxNumberOfAccessTokens;
+
+    @Value("${dat.uuid.revoke_other_tokens:true}")
+    private boolean revokeOtherTokens;
+
+    private static final Log log = LogFactory.getLog(UuidDataAccessTokenServiceImpl.class);
+
+    @Override
+    public DataAccessToken createDataAccessToken(String username) {
+        return createDataAccessToken(username, revokeOtherTokens);
+    }
+
+    // create a data access token (randomly generated UUID) and insert corresponding record into table with parts:
+    // username
+    // uuid
+    // expiration date (current time + 1 month)
+    @Override
+    public DataAccessToken createDataAccessToken(String username, boolean allowRevocationOfOtherTokens) {
+        if (username == null || username.trim().length() == 0) {
+            throw new IllegalArgumentException("username cannot be empty");
+        }
+        if (getNumberOfTokensForUsername(username) >= maxNumberOfAccessTokens) {
+            if (allowRevocationOfOtherTokens) {
+                revokeOldestDataAccessTokenForUsername(username);
+            } else {
+                throw new MaxNumberTokensExceededException("User has reached max number of tokens allowed (" + maxNumberOfAccessTokens + "). An existing token must expire or be revoked before another token can be assigned.");
+            }
+        }
+        String uuid = UUID.randomUUID().toString();
+        Calendar calendar = Calendar.getInstance();
+        Date creationDate = calendar.getTime();
+        calendar.add(Calendar.SECOND, datTtlSeconds);
+        Date expirationDate = calendar.getTime();
+
+        DataAccessToken dataAccessToken = new DataAccessToken(uuid, username, expirationDate, creationDate);
+        dataAccessTokenRepository.addDataAccessToken(dataAccessToken);
+        return dataAccessToken;
+    }
+
+    // get all user tokens/uuids sorted from oldest to newest
+    @Override
+    public List<DataAccessToken> getAllDataAccessTokens(String username) {
+        List<DataAccessToken> allDataAccessTokens = dataAccessTokenRepository.getAllDataAccessTokensForUsername(username);
```

Comment:
@khzhu the tokens are self-authenticating and cannot be used to get access to a study that a user does not already have access for in the portal. 

In portal instances where tokens are not supported (such as the public portal), the user would receive a message stating that they do not have access to this resource (meaning token authentication). We also hide the ability to generate tokens from users when the deployed portal instance doesn't require any sort of authentication.

Reply:
@sheridancbio can you comment on this also?

Reply:
This is the service layer. If the web layer has called this function, it means that the portal is running in a mode that supports using UUID tokens (because this token service (the UuidDataAccessTokenService) has been configured to be used).
If the portal is installed properly (the database exists and the database username and password have been correctly entered into the portal.properties file) then I don't think any exceptions should be expected from the service layer here. There are a small set of users who are prohibited from using tokens, but these users are stopped at the web layer (see private Set<String> usersWhoCannotUseTokenSet; in DataAccessTokenController.java. If any prohibited user tries to access the web API for data access tokens, there is a call to function getAuthenticatedUser(), and there they will receive a DataAccessTokenProhibitedUserException instead of calling the service layer. Users who have not logged in also are filtered out there with DataAccessTokenNoUserIdentityException. So once the service layer call is made, the user is both logged in and is also allowed to use data access tokens.

Reply:
Exceptions already handled by the control layer, make sense.